File: examples/typescript/vault-access/src/access.ts
// copyright © 2025-2026 bkey inc. all rights reserved.

/**
 * TypeScript vault-access example.
 *
 * Shows both patterns side-by-side:
 *
 *   1. Generic one-line CIBA approval — `bkey.approve(...)` returns an
 *      EdDSA-signed JWT after the user approves on their phone.
 *
 *   2. End-to-end encrypted vault access — ephemeral X25519 keypair,
 *      phone encrypts with X25519 ECDH + AES-256-GCM, you decrypt locally.
 *      The backend never sees plaintext.
 *
 * See docs/guides/encryption.mdx for the full envelope layout.
 *
 * Run:
 *   npm run dev
 */

import 'dotenv/config';
import { createDecipheriv, createHash } from 'node:crypto';
import { BKey, pollAccessRequest } from '@bkey/sdk';
import { x25519 } from '@noble/curves/ed25519';

const BKEY_API_URL = process.env.BKEY_API_URL ?? 'https://api.bkey.id';
const BKEY_CLIENT_ID = process.env.BKEY_CLIENT_ID;
const BKEY_CLIENT_SECRET = process.env.BKEY_CLIENT_SECRET;
const BKEY_USER_DID = process.env.BKEY_USER_DID;
const ITEM_NAME = process.env.VAULT_ITEM_NAME ?? 'openai-api-key';

if (!BKEY_CLIENT_ID || !BKEY_CLIENT_SECRET || !BKEY_USER_DID) {
  console.error(
    'Missing credentials. Set BKEY_CLIENT_ID, BKEY_CLIENT_SECRET, ' +
      'BKEY_USER_DID in .env (copy from .env.example).',
  );
  process.exit(1);
}

const bkey = new BKey({
  apiUrl: BKEY_API_URL,
  clientId: BKEY_CLIENT_ID,
  clientSecret: BKEY_CLIENT_SECRET,
  did: BKEY_USER_DID,
});

/**
 * Decrypt the E2EE ciphertext returned by the phone.
 *
 * Envelope layout (base64-decoded):
 *   phonePubKey (32 bytes) || iv (12 bytes) || authTag (16 bytes) || ciphertext
 *
 * Derivation:
 *   sharedSecret = X25519(ourEphemeralPriv, phonePubKey)
 *   aesKey       = SHA-256(sharedSecret)
 *   plaintext    = AES-256-GCM_decrypt(aesKey, iv, ciphertext, authTag)
 */
function decryptE2EE(e2eeCiphertext: string, ephemeralPrivateKey: Uint8Array): string {
  const buf = Buffer.from(e2eeCiphertext, 'base64');
  const phonePubKey = buf.subarray(0, 32);
  const iv = buf.subarray(32, 44);
  const authTag = buf.subarray(44, 60);
  const ciphertext = buf.subarray(60);

  const sharedSecret = x25519.getSharedSecret(ephemeralPrivateKey, phonePubKey);
  const aesKey = createHash('sha256').update(sharedSecret).digest();

  const decipher = createDecipheriv('aes-256-gcm', aesKey, iv);
  decipher.setAuthTag(authTag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString('utf8');
}

async function demoGenericApproval(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Part 1: Generic biometric approval');
  console.log('='.repeat(60));
  console.log("Check your phone — Face ID prompt incoming.\n");

  const result = await bkey.approve(`Read ${ITEM_NAME} for one API call`, {
    scope: 'approve:action',
  });

  if (!result.approved) {
    console.log('Denied on device. Aborting.');
    process.exit(1);
  }

  console.log(`Approved. JWT prefix: ${result.accessToken.slice(0, 24)}...\n`);
}

async function demoVaultAccess(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Part 2: End-to-end encrypted vault access');
  console.log('='.repeat(60));

  // 1. Generate ephemeral X25519 keypair for this request.
  //    The private key never leaves this process.
  const ephemeralPrivateKey = x25519.utils.randomPrivateKey();
  const ephemeralPublicKey = x25519.getPublicKey(ephemeralPrivateKey);

  // 2. Create the access request.
  const access = (await bkey.createAccessRequest({
    itemName: ITEM_NAME,
    fieldPath: 'value',
    purpose: 'TypeScript example: one-shot API call',
    ephemeralPublicKey: Buffer.from(ephemeralPublicKey).toString('base64'),
  })) as { id: string };

  console.log(`Access request created: ${access.id}`);
  console.log("Approve on your phone. It'll show the item name + purpose.\n");

  // 3. Poll until approved / denied / timeout. The backend re-forwards the
  //    phone-produced ciphertext; it can't read it.
  const status = await pollAccessRequest(bkey, access.id, 120_000);

  if (!status.e2eeCiphertext) {
    console.error('Approved but no ciphertext returned. Status:', status.status);
    process.exit(1);
  }

  // 4. Decrypt locally with our ephemeral private key.
  const plaintext = decryptE2EE(status.e2eeCiphertext, ephemeralPrivateKey);

  console.log('Decrypted successfully:');
  console.log(`  prefix: ${plaintext.slice(0, 6)}...`);
  console.log(`  length: ${plaintext.length} chars`);
  console.log('\nUse `plaintext` in-memory; do not log or persist it.');
}

async function main(): Promise<void> {
  try {
    await demoGenericApproval();
    await demoVaultAccess();
  } catch (err) {
    console.error('Example failed:', err instanceof Error ? err.message : err);
    process.exit(1);
  }
}

void main();

File: examples/python/vault-access/access.py
"""Python vault-access example.

Shows how a Python program reads an encrypted secret from the BKey vault
and uses it, after the user approves on their phone with Face ID.

Two patterns side-by-side:

  1. Generic one-line CIBA approval — the universal pattern for any
     sensitive action. `client.approve(...)` returns a signed JWT proving
     the user consented.

  2. Vault access via `bkey wrap` — the recommended way to pull an E2EE
     secret into your Python process. The CLI handles X25519 ECDH and
     AES-256-GCM decryption so you don't have to.

Prereqs:
  - Install the BKey CLI: npm install -g @bkey/cli
  - Create agent credentials: bkey auth login && bkey auth setup-agent --save
  - Store a test secret: bkey vault store openai-api-key --field value=sk-test-...
  - Copy .env.example → .env and fill it in.

Run:
  python access.py
"""

import os
import subprocess
import sys
from pathlib import Path

from bkey import BKeyClient
from bkey.exceptions import ApprovalDeniedError, ApprovalTimeoutError

ITEM_NAME = "openai-api-key"


def _load_env() -> None:
    env_path = Path(__file__).with_name(".env")
    if not env_path.exists():
        return
    for line in env_path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def demo_generic_approval(client: BKeyClient, user_did: str) -> None:
    print("=" * 60)
    print("Part 1: Generic biometric approval")
    print("=" * 60)
    print("Approve the prompt on the user's phone (Face ID)...")
    try:
        result = client.approve(
            message="Python example — read a vault secret",
            user_did=user_did,
            scope="approve:action",
        )
    except ApprovalDeniedError:
        print("Denied on device. Aborting.")
        sys.exit(1)
    except ApprovalTimeoutError:
        print("Timed out waiting for approval.")
        sys.exit(1)

    token_preview = (result.access_token or "")[:24]
    print(f"Approved. JWT starts with: {token_preview}...\n")


def demo_vault_access_via_cli() -> None:
    print("=" * 60)
    print("Part 2: Vault access (E2EE via `bkey wrap`)")
    print("=" * 60)
    print(f'Reading vault item "{ITEM_NAME}" — phone will prompt for approval.')
    print("`bkey wrap` handles X25519 ECDH + AES-256-GCM decryption, then")
    print("injects the plaintext value into this subprocess as an env var.\n")

    inner_script = (
        "import os;"
        "key = os.environ.get('OPENAI_API_KEY', '');"
        "print(f'Key loaded (len={len(key)}, prefix={key[:5]})') "
        "if key else print('Key missing')"
    )
    cmd = [
        "bkey", "wrap",
        "--env", f"OPENAI_API_KEY={{vault:{ITEM_NAME}}}",
        "--purpose", "Python example: read OpenAI key",
        "--",
        sys.executable, "-c", inner_script,
    ]

    try:
        subprocess.run(cmd, check=True)
    except FileNotFoundError:
        print(
            "`bkey` CLI not found on PATH. Install it: `npm install -g @bkey/cli`."
        )
    except subprocess.CalledProcessError as err:
        print(f"Vault access failed (exit {err.returncode}).")


def main() -> None:
    _load_env()

    client_id = os.environ.get("BKEY_CLIENT_ID")
    client_secret = os.environ.get("BKEY_CLIENT_SECRET")
    user_did = os.environ.get("BKEY_USER_DID")
    base_url = os.environ.get("BKEY_API_URL", "https://api.bkey.id")

    if not (client_id and client_secret and user_did):
        print(
            "Missing credentials. Set BKEY_CLIENT_ID, BKEY_CLIENT_SECRET, "
            "BKEY_USER_DID in .env (copy from .env.example)."
        )
        sys.exit(1)

    client = BKeyClient(
        client_id=client_id,
        client_secret=client_secret,
        base_url=base_url,
    )

    demo_generic_approval(client, user_did)
    demo_vault_access_via_cli()


if __name__ == "__main__":
    main()
